src/engine/homeds/btree/btree.hpp

```diff
@@ -735,7 +757,192 @@ class Btree {
+                                                         LOCKTYPE_READ, nullptr);
+                        unlock_node(m_next_left_node, LOCKTYPE_READ);
+
+                        if (ret != btree_status_t::success) {
+                            LOGERROR("Cannot read sibling node for {}", m_next_left_node);
+                            resp.json["error"] = "Cannot read sibling node for {}" + m_next_left_node->to_string();
+                            return resp;
+                        }
+
+                        if (next_node->is_leaf() == false) {
+                            LOGERROR("next_node: {} is not leaf, not expected!", next_node->to_string());
+                            resp.json["error"] = "node: " + next_node->to_string() + " is not leaf, internal error. ";
+                            return resp;
+                        }
+
+                        m_next_left_node = next_node;
+                    }
+                }
+            } else {
+                // if we are here, it means it is a continue of next pagination,
+                // let's lock the next left node firstly;
+                lock_and_refresh_node(m_next_left_node, LOCKTYPE_READ, nullptr);
+            }
+
+            //
+            // fall through
+            //
+        }
+
+        // both range query and/or pagination starts here;
+
+        // m_next_left_node is in locked state now;
+        auto cur_node = m_next_left_node;
+        bool has_more = true;
+        for (int i = 0u; i < request.batch_size; ++i) {
+            // fill the response
+            std::string node_str = "leaf_" + std::to_string(i);
+            resp.json[node_str] = cur_node->to_string();
+
+            // get next node in chain;
+            if (cur_node->get_next_bnode() == empty_bnodeid) {
+                // no more leaf nodes, let's break out;
+                has_more = false;
+                break;
+            }
+
+            BtreeNodePtr next_node = nullptr;
+            auto ret =
+                read_and_lock_sibling(cur_node->get_next_bnode(), next_node, LOCKTYPE_READ, LOCKTYPE_READ, nullptr);
+            unlock_node(cur_node, LOCKTYPE_READ);
+
+            HS_DBG_ASSERT_EQ(ret, btree_status_t::success);
+            if (ret != btree_status_t::success) {
+                LOGERROR("Cannot read sibling node for {}", cur_node);
+                resp.json["error"] = "Cannot read sibling node for {}" + cur_node->to_string();
+                return resp;
+            }
+
+            HS_DBG_ASSERT_EQ(next_node->is_leaf(), true);
+            cur_node = next_node;
+        }
+
+        unlock_node(cur_node, LOCKTYPE_READ);
+
+        if (has_more) {
+            m_next_left_node = cur_node;
+            // remember where we are, this cursor also tells us whether this
+            // is a range query;
+            resp.json["next_cursor"] = m_next_left_node->get_node_id();
+            resp.json["has_more"] = "true";
```

Comment:
We can use next_cursor is present or not to know if more data is available or not. This is same for all mdouels.